### PR TITLE
Add support for sending an event in a "Dry Run" 

### PR DESCRIPTION
**Pull request by knu**
This has long been a missing feature, finally implemented.

### CODE REVIEW

File: app/assets/javascripts/components/utils.js.coffee [hunk 1]
```diff
@@ -34,10 +34,62 @@ class @Utils
...
+            alert 'Invalid JSON object.'
+            return
+          if json == ''
+            if with_event is 'yes'
+              alert 'Event is required for this agent to run.'
+              return
+            dry_run_data = data
+          else
+            dry_run_data = "event=#{encodeURIComponent(json)}&#{data}"
+          $(body).closest('[role=dialog]').on 'hidden.bs.modal', =>
```
Comment (cantino) [inline app/assets/javascripts/components/utils.js.coffee]: Maybe `showDynamicModal` should call `body` with a callback hash or something?

File: app/concerns/dry_runnable.rb [hunk 2]
```diff
@@ -14,7 +14,12 @@ class << self
 
     begin
       raise "#{short_type} does not support dry-run" unless can_dry_run?
-      check
+      if event
+        raise "This agent cannot receive an event!" unless can_receive_events?
+        receive([event])
```
Comment (cantino) [inline app/concerns/dry_runnable.rb]: Some agents can both `check` and `receive` events, like the website agent.  I guess specifying the event in the UI is how you select which mode to dry run?

File: app/controllers/agents_controller.rb [hunk 3]
```diff
@@ -50,7 +50,12 @@ def dry_run
     agent.name ||= '(Untitled)'
 
     if agent.valid?
-      results = agent.dry_run!
+      if event_payload = params[:event]
+        dummy_agent = Agents::ManualEventAgent.new(name: 'Dry-Runner')
+        event = dummy_agent.events.build(user: current_user, payload: event_payload)
```
Comment (cantino) [inline app/controllers/agents_controller.rb]: Might be worth calling `readonly!` on this?

File: app/concerns/dry_runnable.rb (hunk 2 above)
Comment (cantino) [inline app/concerns/dry_runnable.rb]: Ah, that's your `maybe` case.  Got it.

File: app/helpers/agent_helper.rb [hunk 4]
```diff
@@ -37,4 +37,16 @@ def agent_controllers(agent, delimiter = ', ')
       }.join(delimiter).html_safe
     end
   end
+
+  def agent_dry_run_with_event(agent)
```
Comment (cantino) [inline app/helpers/agent_helper.rb]: Maybe `agent_dry_run_with_event_mode` ?

File: spec/controllers/agents_controller_spec.rb [hunk 5]
```diff
@@ -377,6 +377,19 @@ def valid_attributes(options = {})
         [users(:bob).agents.count, users(:bob).events.count, users(:bob).logs.count, agent.name, agent.updated_at]
       }
     end
+
+    it "accepts an event" do
```
Comment (cantino) [inline spec/controllers/agents_controller_spec.rb]: Should there be a spec on dry_runnable.rb too?

**Comment by cantino** [comment]
Awesome!  I'll play with this.  Great idea.

File: app/controllers/agents_controller.rb (hunk 3 above)
Comment (knu) [inline app/controllers/agents_controller.rb]: Good point!  I know `event` will never be saved before `dummy_agent` can be saved, which cannot when it is missing `user`, but I'll make that clear.

File: app/helpers/agent_helper.rb (hunk 4 above)
Comment (knu) [inline app/helpers/agent_helper.rb]: Renamed!

File: spec/controllers/agents_controller_spec.rb (hunk 5 above)
Comment (knu) [inline spec/controllers/agents_controller_spec.rb]: Added!

File: app/assets/javascripts/components/utils.js.coffee (hunk 1 above)
Comment (knu) [inline app/assets/javascripts/components/utils.js.coffee]: Could you elaborate?

Comment (cantino) [inline app/assets/javascripts/components/utils.js.coffee]: My understanding is that you're doing this to detect when the modal is closed, right?  Could we use the `onHide` callback of `showDynamicModal` instead?

**Comment by cantino** [comment]
Works well!  It'd be cool if re-running the Dry Run would default to reusing the previous run's Event JSON.

**Comment by knu** [comment]
Nice idea!  I implemented it!

**Pull request merged**